Start of Kobj support: Semaphores and Threads

d3zd3z opened the pull request:
This PR pulls out some of what is implemented in #5, to try to reduce how much code needs to be reviewed.

In addition, this also incorporates @teburd suggestions around the safety of `Thread` and related.

The thread initialization is now safe in the sense that a given thread can only be initialized once, and a given stack can also only be used once.

The `sys::Semaphore` is safe to get and use, based on the documentation of the zephyr API.

A disadvantage here is that the initialization is more complicated, ~~and currently stacks cannot be declared in `kobj_define!` as an array. A possible solution to this is with either a proc-macro, or to find a supporting proc macro package to help with this~~ (fixed).

cfriedt commented on zephyr/src/object.rs:
```diff
@@ -0,0 +1,231 @@
+where
+    StaticKernelObject<T>: Wrapped,
+{
+    /// Construct an empty of these objects, with the zephyr data zero-filled.  This is safe in the
+    /// sense that Zephyr we track the initialization, and they start in the uninitialized state.
```
I appreciate the attention to good docs.

Not sure if these suggestions are the same as your original intent.
```suggestion
    /// Construct an empty instance of these objects, with the zephyr data zero-filled.  This is safe in the
    /// sense that in Zephyr we track the initialization, and they start in the uninitialized state.
```

d3zd3z replied:
I believe this should be addressed, namely by rewriting the whole doc section.

cfriedt commented on zephyr/src/sys/sync.rs:
```diff
@@ -0,0 +1,140 @@
+    /// permitted count.
+    pub fn give(&self) {
+        unsafe {
+            k_sem_give(self.item)
```
I needed to remind myself that omitting the semicolon with the last expression of a function returns that value 🦀

Does the documentation engine publish return type information for these functions?

d3zd3z replied:
It would have to be declared in the code, it is never inferred. In this case, give returns nothing, and the zephyr function is void.

cfriedt reviewed (commented):
Looks reasonable to me - just a few comments.

d3zd3z commented:
> Looks reasonable to me - just a few comments.

A couple of things I’d like to do are to add some more comments justifying the safety of code that uses unsafe, see if I can make a few pub fields not pub with constructors, and try to figure out how to allow declaring an array of stacks.

d3zd3z commented:
I have implemented `kobj_define!` support for arrays of thread stacks, reflecting this in the philosopher docs.

I've also added a docgen, which can be built, and then allows `cargo doc` to generate the documentation. This will reveal any warnings about broken links, and also have a local rendered copy. I've fixed the few warnings that crept in because the docs weren't being generated.

cfriedt commented on samples/philosophers/src/lib.rs:
```diff
@@ -0,0 +1,134 @@
+    Duration::millis_at_least(((delay + 1) * period) as Tick)
+}
+
+kobj_define! {
```
This looks good - certainly an improvement

Ablu commented on zephyr/src/object.rs:
```diff
@@ -0,0 +1,148 @@
+pub struct StaticKernelObject<T> {
+    #[allow(dead_code)]
+    /// The underlying zephyr kernel object.
+    value: UnsafeCell<T>,
```
I may not have grasped the full context yet, but to me this mostly seems to be used as a `MaybeUninit`? :thinking:  Could that be used instead here? It reads like it will do the same job, but provide some more explicit meaning.

Also: Nothing prevents somebody from moving such an object, right? Are those C counterparts fine with being moved? I would assume synchronization primitives keep all kind of self referential pointers internally. So I am a bit surprised that this can safely be wrapped without some sort of Pinning / modeling things are references to ensure things cannot be moved around.

d3zd3z replied:
The `UnsafeCell` is absolutely critical to this working. It is treated as magical by the Rust compiler and basically tells it that even though the container isn't `mut`, this `T` inside of it might change. Since the C code is likely to change these, we need to declare this like this.

Ablu commented on zephyr/src/object.rs:
```diff
@@ -0,0 +1,148 @@
+        }
+        let result = self.get_wrapped();
+        self.init.store(KOBJ_INITIALIZED, Ordering::Release);
+        Some(result)
```
Is there an example that uses this (maybe my grep skills fail me :P)?

d3zd3z replied:
Look in samples. It is used by any user code using statically declared kernel objects.

The reason to have these two types is that most kernel objects in Zephyr need to be declared statically. But static globals aren't very useful in Rust (without unsafe), so we have this `init_once` method that is called on the global static to return a single instance that wraps the pointer to the static, and can be used more freely in Rust.

Ablu commented on zephyr/src/sys/sync.rs:
```diff
@@ -0,0 +1,124 @@
+    time::Timeout,
+};
+
+/// A zephyr `k_sem` usable from safe Rust code.
+#[derive(Clone)]
+pub struct Semaphore {
+    /// The raw Zephyr `k_sem`.
+    item: *mut k_sem,
+}
```
This seems to work fine as other code carefully avoids constructing this for non-static semahpores. Does zephyr also have dynamic ones? I found these internal pointers made it harder to reason about safety eventually.

I wonder: Have you considered something like:
```
[repr(transparent)]
pub struct Semaphore { 
    value: UnsafeCell<MaybeUninit<k_sem>>,
    _pin: PhantomPinned,
}
```
?
(see https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/rust/kernel/types.rs#n261 / https://github.com/rust-lang/rust/issues/43467 for context)

Then of course most code would work with a `&Semaphore`, but that seems like a more direct mapping of C -> Rust to me?

If all you ever deal with are static objects, then I guess this is fine too. But I think the direct wrapping is going to be a pattern elsewhere eventually anyway? It seems to me like that may also allow to unify this together with the "static object" that currently needs to be defined separately?

d3zd3z replied:
So `Semaphore` is _not_ a static semaphore, it is a singleton reference to some kind of semaphore. The StaticSemaphore is one way of declaring the semaphores (and currently the only way).

I intend to add dynamic variants of many of these, but as the Zephyr dynamic object support is rather poor at the moment, these will likely just have a pool of the objects that they take from. The `Semaphore` type will be the same, but there will lookly be a `Semaphore::new` that returns one from the pool. When it is dropped, it will be returned to the pool.

Ablu reviewed (commented):
Cool stuff! :tada: 

I did not manage to get through all commits today. But here are some comments already :).
There is a decent chance that some of my confusion is due to my limited knowledge about Zephyr internals!

It is mostly the different types for "StaticFoo" and "Foo" that confuse me a bit... My naive feeling would be that there should be ways to do a 1:1 mapping of C type to Rust type without needing the two variants? But again, chances are I miss things! :innocent:

teburd reviewed (approved):
Seems ok to me, gotta start somewhere with this stuff and I think my major concerns are mostly covered. Time will tell what ends up working/not working more definitively.

d3zd3z commented on zephyr-build/src/lib.rs:
```diff
@@ -54,24 +54,22 @@ pub fn build_kconfig_mod() {
     for line in BufReader::new(file).lines() {
         let line = line.expect("reading line from dotconfig");
         if let Some(caps) = config_hex.captures(&line) {
-            writeln!(&mut f, "    #[allow(dead_code)]").unwrap();
-            writeln!(&mut f, "    pub const {}: usize = {};",
+            writeln!(&mut f, "#[allow(dead_code)]").unwrap();
+            writeln!(&mut f, "pub const {}: usize = {};",
                 &caps[1], &caps[2]).unwrap();
         } else if let Some(caps) = config_int.captures(&line) {
-            writeln!(&mut f, "    #[allow(dead_code)]").unwrap();
-            writeln!(&mut f, "    pub const {}: isize = {};",
+            writeln!(&mut f, "#[allow(dead_code)]").unwrap();
+            writeln!(&mut f, "pub const {}: isize = {};",
                 &caps[1], &caps[2]).unwrap();
         } else if let Some(caps) = config_str.captures(&line) {
-            writeln!(&mut f, "    #[allow(dead_code)]").unwrap();
-            writeln!(&mut f, "    pub const {}: &'static str = {};",
+            writeln!(&mut f, "#[allow(dead_code)]").unwrap();
+            writeln!(&mut f, "pub const {}: &'static str = {};",
                 &caps[1], &caps[2]).unwrap();
         }
     }
-    writeln!(&mut f, "}}").unwrap();
```
It didn't sneak, it is very intentional. I needed to move the `pub mod kconfig {` up into `lib.rs` so that I could add both doc comments and directives to not fail on missing docs on the entries. To do that, I had to remove that from the generated code, and I unindented it as well.

d3zd3z commented on zephyr/src/error.rs:
```diff
@@ -0,0 +1,56 @@
+/// Wraps a value with a possible Zephyr error.
+pub type Result<T> = core::result::Result<T, Error>;
+
+/// Map a return result from Zephyr into an Result.
+///
+/// Negative return results being considered errors.
+pub fn to_result(code: c_int) -> Result<c_int> {
+    if code < 0 {
+        Err(Error(-code as u32))
+    } else {
+        Ok(code)
+    }
+}
+
+/// Map a return result, with a void result.
+pub fn to_result_void(code: c_int) -> Result<()> {
+    to_result(code).map(|_| ())
+}
```
Definitely a consideration. I still think they are useful for an app user that wants to use a binding that we haven't wrapped yet.

otavio replied:
Reducing the API may make it easier for people to use. This extra API seems to be a convenience API, but, at least for me, it just adds extra API for people who will use the crate to learn.

One possibility here would be to add a `ZephyrCode` type that could implement those methods. I think using the type `Error` is misleading because we are using it for error and also for normal returning code.

d3zd3z commented on zephyr/src/object.rs:
```diff
@@ -0,0 +1,148 @@
+    /// sense that Zephyr we track the initialization, and they start in the uninitialized state.
+    pub const fn new() -> StaticKernelObject<T> {
+        StaticKernelObject {
+            value: unsafe { mem::zeroed() },
```
Honestly, I think this whole method should probably just be unsafe.

d3zd3z commented on zephyr/src/object.rs:
```diff
@@ -0,0 +1,148 @@
+            Ordering::Acquire)
+        {
+            return None;
+        }
```
I'll look at it.

d3zd3z commented on zephyr/src/object.rs:
```diff
@@ -0,0 +1,148 @@
+    type T;
+
+    /// Initialize this kernel object, and return the wrapped pointer.
+    fn get_wrapped(&self) -> Self::T;
```
They don't follow the semantics of Rust references, they are never dereferenced from Rust, and are really just part of the C api. Also, this has been remove anyway (but it is still a wrapped pointer).

d3zd3z commented:
@Ablu I think I'm going to move forward with merging this, and address some of your concerns in the next PR. The ones about code organization and safety comments don't really affect the code itself. Hopefully, the concerns about the static stuff will make more sense in the next PR as I add `sys::Mutex` and the more Rust-friendly `sync::Mutex` that makes use of it. I chose Semaphore for this PR because `k_sem` is about the only kernel object in Zephyr that is perfectly usable directly from safe Rust code. The rest will have sys wrappers, but will have other mechanisms to make them more usable from safe code.
